### ISSUE TITLE
Infinite while loop in rotated bounding box detector

### ISSUE THREAD

**Issue by edwardchaos** [facebookresearch/detectron2#339]
Model for rotated bounding box detection trains for many iterations until it gets stuck in an infinite loop.

Infinite while loop experienced in file box_regression.py apply_deltas for class Box2BoxTransformRotated line 216.
        
while len(torch.where(pred_angle < -180.0)[0]) > 0:
    pred_angle[torch.where(pred_angle < -180.0)] += 360.0

## Expected behavior
Loop should change all angles in range -180, 180 but loops forever

------------------------  --------------------------------------------------
sys.platform              linux
Python                    3.7.5 (default, Oct 25 2019, 15:51:11) [GCC 7.3.0]
Numpy                     1.17.3
Detectron2 Compiler       GCC 7.4
Detectron2 CUDA Compiler  10.1
DETECTRON2_ENV_MODULE     <not set>
PyTorch                   1.3.1
PyTorch Debug Build       False
torchvision               0.4.2
CUDA available            True
GPU 0,1                   GeForce RTX 2080 Ti
CUDA_HOME                 /usr/local/cuda-10.1
NVCC                      Cuda compilation tools, release 10.1, V10.1.243
Pillow                    6.2.1
cv2                       4.1.1
------------------------  --------------------------------------------------
PyTorch built with:
  - GCC 7.3
  - Intel(R) Math Kernel Library Version 2019.0.4 Product Build 20190411 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.20.5 (Git Hash 0125f28c61c1f822fd48570b4c1066f96fcb9b2e)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CUDA Runtime 10.1
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_37,code=compute_37
  - CuDNN 7.6.3
  - Magma 2.5.1
  - Build settings: BLAS=MKL, BUILD_NAMEDTENSOR=OFF, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Wno-stringop-overflow, DISABLE_NUMA=1, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=True, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_STATIC_DISPATCH=OFF,

**Comment by edwardchaos** [comment]
Just checked what the angle of pred_angle is, it's something like
tensor([-35781664.0000,  14855407.0000,  45173272.0000,  ...,
          2197290.5000,  -4153832.5000,  -2481959.5000], device='cuda:0')
and has size torch.Size([245760]) so it takes forever.

**Comment by edwardchaos** [comment]
How about instead of adding / subtracting only 360 at a time, add / subtract a multiple of 360 depending on the value?

**Comment by edwardchaos** [comment]
angle deltas produced by rpn, I'm suspecting this is a result of exploding gradients

**Comment by ppwwyyxx** [comment]
It does feel like the angles might be exploding to `NaN` of `Inf`, otherwise the loop is supposed to end. I think we should add an assertion at this place and abort training when abnormal values are found.

**Comment by edwardchaos** [comment]
I added gradient clipping, network is training fine now.

**Comment by Samjith888** [comment]
> I added gradient clipping, network is training fine now.

Where you added the gradient clipping?

**Comment by edwardchaos** [comment]
@Samjith888 probably not the cleanest solution, but I clipped at the end

**Comment by Samjith888** [comment]
> @Samjith888 probably not the cleanest solution, but I clipped at the end
Im my case , train loss reach upto 0.561 .. after that its remains the same .. so i tried to changing the base learning rate . But nothing worked out.. I thought may be gradient clipping may be a good solution..
